ggjdev/php-developer-tech-test
# Fix issues in the company search form

## Description of the problem
There are some problems with the newly implement search form that will allow customers to search for companies to carry out surveys, this PR fixes these issues and gets the form working as expected.

## Assumptions
I have made the following assumptions in this PR:
- The maximum number of matches is controller by an environment variable: this is because there was a variable in the example environment file called `MAX_MATCHED_COMPANIES` that was set to `3` - the expected maximum number of returned results.
- The pick method is not needed: looking at the code, it seemed logical to specify the number of matches to find when doing the matching. It reduces the amount of processing that needs to be done in PHP and simplifies the PHP code.
- The search will not be looking companies that survey more than 5 bedrooms. Looking at the way the number of bedrooms companies survey is stored in the database, searching for bedrooms greater than a number would be difficult, limited and not very efficient. It would need a larger separate task to tackle this, so for the time being referenced to greater than 5 were removed.

## Changes in this PR
- Add the composer lock file to ensure that all environments including development and production are using the same versions of dependencies, it is recommended in the [composer documentation](https://getcomposer.org/doc/01-basic-usage.md#commit-your-composer-lock-file-to-version-control).
- Make the form controller extend the base controller to fix the `render` method not existing on the `FormController`.
- Define the parent template blocks that are being over-ridden in the views as described [here in the twig documentation](https://twig.symfony.com/doc/3.x/templates.html#template-inheritance).
- Register the POST route for form submission that calls the `submit` function in the `FormController`.
- Add a function to fetch the max matches from the ENV - storing this in configuration removes the need to create a branch and go through a development process to change the number of items in search results.
- Added a function to fetch the postcode prefix - it felt best to extract this logic into its own function.
- Add the logic to get the match function to find companies that match the search
- Add logic to deduct credits from the matched companies
- Fix passing the results into the results view
- Randomly select matched companies. The approach taken me run into some scaling issues down the line, this should be fine for a while especially considering there are constraints on the results and indexes could be applied to improve this first. Here is some more information about selecting random rows from the database: https://www.basedash.com/blog/random-select-in-mysql
- Fixed some accessibility issues found by a quick accessibility test.
- Change the "5 or more bedrooms" option to 5 - as explained in the assumptions.
- Add some basic form validation - at the moment it does not pre-fill what the user selected but this would be a good user experience improvement we could make in the future.

## Testing instructions
1. Check out this branch.
2. Follow the instructions in the README to get your environment working.
3. Open a web browser and go to your local development URL followed by `/form`.
4. Fill out the form and submit.
5. Check that the responses are expected.
6. Go to the database and check the credits for the returned companies.
7. Refresh the page and resubmit the form prompt.
8. Check the DB to see if the returned results have had their credits decremented.
9. Test around and try to cover the "sad paths".
